providers/implementations/encode_decode/decode_lmsxdr2key.c

```diff
@@ -0,0 +1,163 @@
+{
+    if (selection == 0)
+        return 1;
+
+    if ((selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0)
+        return 1;
+
+    return 0;
+}
+
+static int lmsxdr2key_decode(void *vctx, OSSL_CORE_BIO *cin, int selection,
+                             OSSL_CALLBACK *data_cb, void *data_cbarg,
+                             OSSL_PASSPHRASE_CALLBACK *pw_cb, void *pw_cbarg)
+{
+    struct lmsxdr2key_ctx_st *ctx = vctx;
+    LMS_KEY *key = NULL;
+    unsigned char buf[LMS_MAX_PUBKEY];
+    size_t length;
+    int ok = 0, inlen;
+    BIO *in;
+
+    in = ossl_bio_new_from_core_bio(ctx->provctx, cin);
+    if (in == NULL)
+        return 0;
+
+    ERR_set_mark();
+    ctx->selection = selection;
+    /* Read the header to determine the size */
+    if (BIO_read(in, buf, 4) != 4)
+        goto next;
+
+    length = ossl_lms_pubkey_length(buf, 4);
+    if (length <= 4)
+        goto next;
+    inlen = (int)length - 4;
+    if (BIO_read(in, buf + 4, inlen) != inlen)
+        goto next;
+    if (selection == 0 || (selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0) {
+        key = ossl_lms_key_new(PROV_LIBCTX_OF(ctx->provctx));
+        if (key == NULL || !ossl_lms_pubkey_decode(buf, length, key)) {
+            ossl_lms_key_free(key);
+            key = NULL;
+        }
+    }
+ next:
+    ERR_clear_last_mark();
+    /*
+     * Indicated that we successfully decoded something, or not at all.
+     * Ending up "empty handed" is not an error.
+     */
+    ok = 1;
+
+    /*
+     * We free resources here so it's not held up during the callback, because
+     * we know the process is recursive and the allocated chunks of memory
+     * add up.
+     */
+    BIO_free(in);
+    in = NULL;
+
+    if (key != NULL) {
+        OSSL_PARAM params[4];
+        int object_type = OSSL_OBJECT_PKEY;
+
+        params[0] =
+            OSSL_PARAM_construct_int(OSSL_OBJECT_PARAM_TYPE, &object_type);
+        params[1] =
+            OSSL_PARAM_construct_utf8_string(OSSL_OBJECT_PARAM_DATA_TYPE,
+                                             (char *)"lms", 0);
+        /* The address of the key becomes the octet string */
+        params[2] =
+            OSSL_PARAM_construct_octet_string(OSSL_OBJECT_PARAM_REFERENCE,
+                                              &key, sizeof(key));
+        params[3] = OSSL_PARAM_construct_end();
+
+        ok = data_cb(params, data_cbarg);
+    }
+
+    BIO_free(in);
+    ossl_lms_key_free(key);
+    return ok;
+}
+
+static int lmsxdr2key_export_object(void *vctx,
+                                    const void *reference, size_t reference_sz,
+                                    OSSL_CALLBACK *export_cb,
+                                    void *export_cbarg)
+{
+    struct lmsxdr2key_ctx_st *ctx = vctx;
+    OSSL_FUNC_keymgmt_export_fn *export =
+        ossl_prov_get_keymgmt_export(ossl_lms_keymgmt_functions);
+    void *keydata;
+
+    if (reference_sz == sizeof(keydata) && export != NULL) {
```

Comment:
If there is a size mismatch we just get a zero return with no indicator as to the issue?
That seems a little sparse for error reporting ...

Reply:
This is the same pattern that is used everywhere.